ygot/render.go

```diff
@@ -177,14 +205,84 @@ func findUpdatedLeaves(leaves map[*path]interface{}, s GoStruct, parentPath []in
 // and populates a gNMI path message with the string components. It should be noted that this
 // functionality does not comply with the gNMI specification, and should be updated in the
 // future.
-// TODO(robjs): Update this functionality based on adoption of the gNMI structured paths
-// when Pictor is ready to support these.
-func interfacePathAsgNMIPath(p []interface{}) *gnmipb.Path {
+func interfacePathAsgNMIPath(p []interface{}, usePathElem bool) (*gnmipb.Path, error) {
 	pfx := &gnmipb.Path{}
 	for _, e := range p {
-		pfx.Element = append(pfx.Element, fmt.Sprintf("%v", e))
+		if usePathElem {
+			pe, ok := e.(*gnmipb.PathElem)
+			if !ok {
+				return nil, fmt.Errorf("invalid path element %v in path %v, expected gnmipb.PathElem, got: %T", e, p, e)
+			}
+			pfx.Elem = append(pfx.Elem, pe)
+		} else {
+			pfx.Element = append(pfx.Element, fmt.Sprintf("%v", e))
+		}
+	}
+	return pfx, nil
+}
+
+// appendgNMIPathElemKey takes an input reflect.Value which must implement KeyHelperGoStruct
+// and appends the keys from it to the last entry in the supplied mapPath, which must be a
+// gNMI PathElem message.
+func appendgNMIPathElemKey(v reflect.Value, mapPath []interface{}) ([]interface{}, error) {
+	if len(mapPath) == 0 {
+		return nil, fmt.Errorf("zero length mapPath supplied", mapPath)
+	}
+	e, ok := mapPath[len(mapPath)-1].(*gnmipb.PathElem)
+	if !ok {
+		return nil, fmt.Errorf("invalid path element %v, not a correct type", mapPath[len(mapPath)-1])
+	}
+	newElem := *e
+
+	if !v.IsValid() || v.IsNil() {
+		return nil, fmt.Errorf("nil value received for element %v", mapPath)
+	}
+
+	gs, ok := v.Interface().(KeyHelperGoStruct)
+	if !ok {
+		return nil, fmt.Errorf("cannot render to gNMI PathElem for structs that do not implement KeyHelperGoStruct, got: %T (%s)", v.Type().Name(), v.Interface())
+	}
+
+	km, err := gs.ΛListKeyMap()
+	if err != nil {
+		return nil, err
+	}
+	newElem.Key = map[string]string{}
+	for kn, k := range km {
+		v, err := keyValueAsString(k)
+		if err != nil {
+			return nil, err
+		}
+		newElem.Key[kn] = v
+	}
+
+	nm := append([]interface{}(nil), mapPath...)
```

Comment:
more idiomatic to use copy here.

Reply:
Sure - this seemed cleaner since for `copy` we need two lines :-)